### PR TITLE
Clean up and refactor `UserAvatar` component

### LINKED ISSUE #63: Cleanup and Refactor `UserAvatar` component
- cleanup imports and remove what's no longer required
- remove all traces of IPFS / `withDataFetcher` logic
- Remove the `Hooked` higher order function
- make sure styles are up to date with the latest ones from the Dapp
- introduce logic to account for thumbnails, as we now have that value in the db

### PR DESCRIPTION
**Pull request by Julianahlk**
## Description

This PR takes care of the cleaning and refactoring of the `UserAvatar` component.

**New stuff** ✨

* Added new lambda function `getReputationForTopDomains `
* Added `TokenInfoPopoverWrapper `
* Created a new  `UserReputationItem` component that handles each item in the user reputation list

**Changes** 🏗

* Cleaned up `UserAvatar` component
* Combined `UserInfoPopover` and `MemberInfoPopover` into one component
* Simplified the `Tag` component

**Deletions** ⚰️

* `InfoPopOver`
* `MemberInfoPopover`
* `UserInfoPopover`

## Testing

1. Add reputation using truffle to a user and a domain to show reputation in the UserInfoPopover:
```c = await IColony.at('0x3a157280ca91bB49dAe3D1619C55Da7F9D4438c3')```
```c.emitDomainReputationReward(1, accounts[0], '1000000000000000000')```

2. Make sure the reputation monitor is ON:
`http://127.0.0.1:3001/reputation/monitor/toggle`
`http://127.0.0.1:3001/reputation/monitor/status`

Without Reputation
Screen Shot 2022-11-23 at 4 58 31 PM

With Reputation
Screen Shot 2022-11-23 at 4 55 10 PM

## TODO

- [x] Cleanup imports and remove what's no longer required
- [x] Remove all traces of IPFS / withDataFetcher logic
- [x] Remove the Hooked higher order function
- [x] Make sure styles are up to date with the latest ones from the Dapp
- [x] Introduce logic to account for thumbnails, as we now have that value in the db
- [x] Clean up, and wire the user tooltip
- [x] Fix the user avatar hover styles
- [x] Create task for the refactoring of getMainClasses in a separate [PR](https://github.com/JoinColony/colonyCDapp/issues/89)

Resolves #63

### CODE REVIEW

**Comment by rdig** [comment]
@Julianahlk 

There's some considerations to be made when you get to the User Tooltip part:
- when getting to the reputation part make sure to  talk with @danbr and @jakubcolony as they already encountered reputation in their respective PRs, so the queries you need are most likely already done
- don't need to implement user's tokens (that logic is not in yet)
- don't need to implement permissions (that will be handled by #41)
- right now all 3 tooltips (user, colony, token) are linked together in a complex component called `InfoPopover`. There's a good argument to be made that the 3 are split and each handled individually, as their requirements are quite complex. Please do so, and only port over / create the logic for the `UserInfo` tooltip component

File: src/redux/transformers/index.ts [hunk 1]
```diff
@@ -125,3 +125,8 @@ export const getAllUserRoles = (colony, userAddress: Address): ColonyRole[] => {
     }, new Set<ColonyRole>()),
   );
 };
+
+export const getUsername = (user?: User) => {
+  if (!user) return '';
+  return user.name;
+};
```
Comment (rdig) [inline src/redux/transformers/index.ts:132]: This is not be needed anymore.

File: src/components/shared/UserAvatar/UserAvatar.tsx [hunk 2]
```diff
@@ -59,71 +56,71 @@ const displayName = 'UserAvatar';
...
+  const username = getUsername(user);
+  const trigger = showInfo ? 'click' : 'disabled';
+  const showArrow = popperOptions && popperOptions.showArrow;
+
+  const renderContent = useMemo(() => {
+    if ('colony' in reputationProps) {
+      const { colony } = reputationProps;
+      if (typeof colony !== 'undefined') {
+        return (
+          <MemberInfoPopover
```
Comment (rdig) [inline src/components/shared/UserAvatar/UserAvatar.tsx:78]: Unless there's something super blocking, `MemberInfoPopover` and `UserInfoPopover` should be just one component

File: src/components/shared/Tag/Tag.tsx [hunk 3]
```diff
@@ -54,6 +54,7 @@ const Tag = ({
         />
       )}
       {text ? (
+        // eslint-disable-next-line react/jsx-no-useless-fragment
```
Comment (rdig) [inline src/components/shared/Tag/Tag.tsx:57]: Why not refactor this to remove the fragment, so that the rule bypass will no longer be required ?

File: src/components/shared/Popover/PopoverWrapper.tsx [hunk 4]
```diff
@@ -80,6 +81,7 @@ const PopoverWrapper = ({
       {popoverContent}
       {state && state.placement && (
         <span
+          // @ts-ignore
```
Comment (rdig) [inline src/components/shared/Popover/PopoverWrapper.tsx:84]: Please investigate and try to fix this type issue

File: src/components/shared/Popover/PopoverWrapper.tsx [hunk 5]
```diff
@@ -63,6 +63,7 @@ const PopoverWrapper = ({
   }, [close, content, contentValues, formatMessage]);
   return (
     <div
+      // @ts-ignore
```
Comment (rdig) [inline src/components/shared/Popover/PopoverWrapper.tsx:66]: Please investigate and try to fix this type issue

File: src/components/shared/InfoPopover/index.ts [hunk 6]
```diff
@@ -1 +1 @@
-export { default, Props } from './InfoPopover';
+export { default, Props } from './TokenInfoPopoverWrapper';
```
Comment (rdig) [inline src/components/shared/InfoPopover/index.ts:1]: Why is the `InfoPopover` index, exporting from `TokenInfoPopover`

**Review by rdig — COMMENTED** [review]
More or less ok, although I would:
- merge `MemberInfoPopover` and `UserInfoPopover`
- Put all separate popovers in their own folder / component, rather than to live under `InfoPopover`

File: src/components/shared/InfoPopover/index.ts (hunk 6 above)
Comment (Julianahlk) [reply src/components/shared/InfoPopover/index.ts]: Not an issue now as I removed the InfoPopover

File: scripts/temp-create-data.js [hunk 7]
```diff
@@ -105,13 +107,15 @@ const createUser = async (username, accountIndex = 0) => {
...
+  const thumbnail = username === 'a' ? thumbnailUserA : username === 'b' ? thumbnailUserB : thumbnailUserC;
+
   const userQuery = await graphqlRequest(
     createUniqueUser,
     {
       input: {
         id: userAddress,
         name: username,
-        profile: { displayName: `User ${username.toUpperCase()}` },
+        profile: { displayName: `User ${username.toUpperCase()}`, thumbnail: thumbnail },
```
Comment (rdig) [inline scripts/temp-create-data.js:118]: You import everything as a single object:
```ts
const userThumbnails = require('./imageData');
```

Then you assign them like this
```suggestion
        profile: { displayName: `User ${username.toUpperCase()}`, thumbnail: userThumbnails[`thumbnailUser${username.toUpperCase()}`] },
```

That way, you won't need the `const` assignment on line 110

File: src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx [hunk 8]
```diff
@@ -1,31 +1,31 @@
...
   isTokenNative: boolean;
 }
 
-const displayName = 'InfoPopover.TokenInfoPopover';
+const displayName = 'TokenInfoPopover';
 
 const TokenInfoPopover = ({ token, isTokenNative }: Props) => {
   return (
-    <>
+    <span>
```
Comment (rdig) [inline src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx:20]: Why was this changed to a `span` element ?

File: src/components/shared/UserAvatar/UserAvatar.tsx [hunk 9]
```diff
@@ -1,129 +1,111 @@
...
+          size={size}
+          title={
+            showInfo ? '' : user?.profile?.displayName || user?.name || address
+          }
+        />
+      </div>
+    </Popover>
   );
+  if (showLink && user) {
+    // Won't this always be lowercase?
```
Comment (rdig) [inline src/components/shared/UserAvatar/UserAvatar.tsx:105]: No it won't

At least currently we don't have any transform that does this. But maybe we should look into this

File: src/components/shared/UserInfoPopover/UserInfoPopover.tsx [hunk 10]
```diff
@@ -0,0 +1,144 @@
+import React from 'react';
+// import isEmpty from 'lodash/isEmpty';
+// import { BigNumber } from 'ethers';
+
+// import { SpinnerLoader } from '~shared/Preloaders';
+import Tag from '~shared/Tag';
+// import {
+//   useColonyNativeTokenQuery,
+//   useUserBalanceWithLockQuery,
+// } from '~data/index';
```
Comment (rdig) [inline src/components/shared/UserInfoPopover/UserInfoPopover.tsx:10]: This can be removed `~data` is not coming back

File: src/components/shared/UserInfoPopover/UserInfoPopover.tsx [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  banned?: boolean;
+}
+
+const displayName = 'UserInfoPopover';
+
+const UserInfoPopover = ({
+  // colony: { colonyAddress },
+  colony,
+  user,
+  banned = true,
```
Comment (rdig) [inline src/components/shared/UserInfoPopover/UserInfoPopover.tsx:35]: Why is this `true` by default? This seems wrong.

**Review by rdig — APPROVED** [review]
Tested and works as expected. I actually expected this not be so functional at it's first review pass, but this is quality work. Very nice @Julianahlk !!! :100: 

Screenshot from 2022-11-24 19-36-02
Screenshot from 2022-11-24 19-36-06
Screenshot from 2022-11-24 19-36-09

Also want to point out, that for some reason you chose to only have 2 out of the 3 avatars as cats :cat: 

Screenshot from 2022-11-24 19-32-40

While I did approve this, please address the code changes / comments I left below, and also, since Jakub's #90 will most likely be merged first, after you rebase, you'll have to re-write the graphql query by hand, since it won't be auto-generated anymore.

After you do the above, please ping me, and I'll re-review.

**Review by jakubcolony — APPROVED** [review]
Awesome work @Julianahlk!

Left a few comments related to GraphQL types but overall everything looks good!

image

File: src/components/shared/Popover/PopoverWrapper.tsx [hunk 12]
```diff
@@ -65,10 +64,14 @@ const PopoverWrapper = ({
     <div
       className={`
         popoverWrapper
-        ${getMainClasses(appearance, styles, {
-          hideArrow: !showArrow,
-          showArrow,
-        })}
+        ${getMainClasses(
+          appearance,
+          styles as unknown as { [k: string]: string },
```
[the remark this answers is not in the document]
Comment (rdig) [reply src/components/shared/Popover/PopoverWrapper.tsx:69]: Fixing `getMainClasses`, we already have a issue for it #89

[the remark this answers is not in the document]
Comment (jakubcolony) [reply src/components/shared/Popover/PopoverWrapper.tsx:69]: 🤦 Cool, thanks!

File: amplify/backend/function/getReputationForTopDomains/src/index.js [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+        ) {
+          return -1;
+        }
+        return 1;
+      })
+      .slice(0, 3);
+
+    return { items: topUserReputations };
+  } catch (error) {
+    return error.toString();
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply amplify/backend/function/getReputationForTopDomains/src/index.js:160]: good catch, I was using that for debugging.

File: scripts/temp-create-data.js (hunk 7 above)
Comment (Julianahlk) [reply scripts/temp-create-data.js:118]: Great suggestion, thanks

File: src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx (hunk 8 above)
Comment (Julianahlk) [reply src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx:20]: What I have noticed is that there are many places in the code that use a jsx conditional statement and are surrounded by either fragments, span or div because these conditional statements cause errors when they are not enclosed in a parent element. When it's a fragment, it gives the "Fragments should contain more than one child" error. I have not found a solution for this yet so I thought a span could work for now. 😅  but I am open to suggestions.

File: src/components/shared/UserAvatar/UserAvatar.tsx (hunk 9 above)
Comment (Julianahlk) [reply src/components/shared/UserAvatar/UserAvatar.tsx:105]: I included the comment from the dapp just incase it was relevant but if if a user name has the possibility of being uppercase it sounds like I should remove the comment and using toLowerCase() should work for now.

File: src/components/shared/UserInfoPopover/UserInfoPopover.tsx (hunk 11 above)
Comment (Julianahlk) [reply src/components/shared/UserInfoPopover/UserInfoPopover.tsx:35]: yes, it should be false, I must have set that to true while testing the Tag component. Nice catch.

**Comment by Julianahlk** [comment]
> Tested and works as expected. I actually expected this not be so functional at it's first review pass, but this is quality work. Very nice @Julianahlk !!! 💯
 
> While I did approve this, please address the code changes / comments I left below, and also, since Jakub's #90 will most likely be merged first, after you rebase, you'll have to re-write the graphql query by hand, since it won't be auto-generated anymore.
> After you do the above, please ping me, and I'll re-review.

Thanks @rdig! I will work on the changes and will ping you again once everything is done. 🙌

> Also want to point out, that for some reason you chose to only have 2 out of the 3 avatars as cats 🐱

Lol 😂 I only liked 2 of the cats that came up on my search. If you wish I can find a third one.

File: amplify/backend/api/colonycdapp/schema.graphql [hunk 14]
```diff
@@ -67,6 +67,21 @@ type TokenFromEverywhereReturn {
...
+  rootHash: String
+}
+
+type UserDomainReputation {
+  domainId: Int!
+  reputationPercentage: String!
+}
+
+type GetReputationForTopDomainsReturn {
+  items: [UserDomainReputation]
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply amplify/backend/api/colonycdapp/schema.graphql]: > Awesome work @Julianahlk!

> Left a few comments related to GraphQL types but overall everything looks good!

Great suggestions @jakubcolony! I implemented them already. 👍

**Comment by Julianahlk** [comment]
> Awesome work @Julianahlk!
> Left a few comments related to GraphQL types but overall everything looks good!

Great suggestions @jakubcolony! I implemented them already. 👍

**Comment by jakubcolony** [comment]
Hi @Julianahlk, my PR for GraphQL typings has now been merged and you can apply those changes to your PR. Let me know if you have any questions

**Comment by Julianahlk** [comment]
> Hi @Julianahlk, my PR for GraphQL typings has now been merged and you can apply those changes to your PR. Let me know if you have any questions

Thanks Jakub! I will be working on that today. 👍

**Comment by Julianahlk** [comment]
@rdig I have implemented the changes and updated the query. Please take a look again when you get a chance.

File: src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx [hunk 15]
```diff
@@ -0,0 +1,58 @@
...
+import { PopperOptions } from 'react-popper-tooltip';
+
+import Popover from '~shared/Popover';
+import { Token } from '~types';
+
+import TokenInfoPopover from './TokenInfoPopover';
+
+interface TokenContentProps {
+  isTokenNative?: boolean;
+  token?: Token;
```
Comment (willm30) [inline src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx:11]: Why is `token` optional here? If `token` is `undefined`, this popover's content is `null`, which means there'd be no reason to render it. Right? This is as much me thinking out loud since I may not be following the code correctly, but if that's correct then lines 36-42 could be removed and this component only rendered if the token prop is defined.

**Review by willm30 — APPROVED** [review]
Nice job! Avatar working nicely.

File: src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx (hunk 8 above)
Comment (Julianahlk) [reply src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx:20]: @rdig just wanted to follow up on this specifically as I keep encountering it and I would like to hear your opinion on the best way to solve this.

File: src/components/shared/UserAvatar/UserAvatar.tsx [hunk 16]
```diff
@@ -1,129 +1,110 @@
...
+    <Popover
+      renderContent={renderContent}
+      popperOptions={popperOptions}
+      trigger={trigger}
+      showArrow={showArrow}
+    >
+      <div
+        className={getMainClasses(
+          {},
+          styles as unknown as { [k: string]: string },
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/shared/UserAvatar/UserAvatar.tsx:84]: This was already discussed above in one of Jakub's comments. There is already several places in the code with this pattern and I think resolving all of them in a separate task would be best.

File: src/components/shared/UserAvatar/UserAvatar.tsx [hunk 17]
```diff
@@ -1,129 +1,110 @@
...
-        showInfo ? '' : user?.profile?.displayName || user?.name || address
-      }
-    />
+  const trigger = showInfo ? 'click' : 'disabled';
+  const showArrow = popperOptions && popperOptions.showArrow;
+  const { profile } = user || {};
+  const imageString = preferThumbnail ? profile?.thumbnail : profile?.avatar;
+
+  const renderContent = useMemo(() => {
+    return <UserInfoPopover colony={colony} user={user} banned={banned} />;
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/shared/UserAvatar/UserAvatar.tsx:71]: You are right, this was leftover from the dapp and seems unnecessary here. Thanks.

File: src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx (hunk 15 above)
Comment (Julianahlk) [reply src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx:11]: The token functionality was not the focus of this PR and I can't test the Token info right now but wouldn't the children inside the popover still get rendered even with null renderContent?

File: src/components/shared/UserInfoPopover/UserReputation.tsx [hunk 18]
```diff
@@ -0,0 +1,136 @@
...
+      );
+      return {
+        ...rest,
+        reputationDomain,
+      };
+    },
+  );
+
+  return (
+    <div className={styles.sectionContainer}>
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/shared/UserInfoPopover/UserReputation.tsx:61]: Ok, for now I extracted the list item to a new component. Maybe you can create separate issues for the other components you would like to see cleaned up? 👍

File: src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx (hunk 8 above)
Comment (jakubcolony) [reply src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx:20]: @Julianahlk I also came across this error and having looked into this, I think it might be best to modify the eslint config for the rule triggering it (as I did in my [extensions PR](https://github.com/JoinColony/colonyCDapp/pull/95/files#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5))

**Review by jakubcolony — APPROVED** [review]
Thanks for making the changes, all looks good! 👍

File: src/components/shared/UserInfoPopover/UserReputation.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (willm30) [reply src/components/shared/UserInfoPopover/UserReputation.tsx:61]: Looks good. Is more of a comment in general that while we're doing the migration and it's fresh in our minds, it's a good chance to refactor some older code.

**Review by danbr — APPROVED** [review]
Looks good.... well done 👍

File: src/components/shared/UserMention/UserMention.tsx [hunk 19]
```diff
@@ -37,21 +39,18 @@ const UserMention = ({
...
-  });
-
-  const { data } = useUserQuery({
-    variables: { address: userAddressData?.userAddress || '' },
-  });
+  const trigger = showInfo ? 'click' : 'disabled';
+  const showArrow = popperOptions && popperOptions.showArrow;
+
+  // const { data: userAddressData } = useUserAddressQuery({
+  //   variables: {
```
Comment (chinins) [inline src/components/shared/UserMention/UserMention.tsx:46]: Why do we have the commented out code?

File: src/redux/transformers/index.ts [hunk 20]
```diff
@@ -38,53 +38,53 @@ const getRolesForUserAndParentDomains = (
...
-  return Array.from(
-    domainRoles
-      .filter(({ domainId }) => domainIds.includes(domainId))
-      .reduce((roleSet, domainRole) => {
-        domainRole.roles.forEach((role) => roleSet.add(role));
-        return roleSet;
-      }, new Set<ColonyRole>()),
-  );
-};
+// const getCombinedRolesForDomains = (
```
Comment (chinins) [inline src/redux/transformers/index.ts:41]: Why is this commented out?

**Review by chinins — APPROVED** [review]
Very nice & clean!

File: src/redux/transformers/index.ts (hunk 20 above)
Comment (Julianahlk) [reply src/redux/transformers/index.ts:41]: The way roles work is supposed to change so commenting it in preparation for that work and as a reminder to not use this for now.

File: src/components/shared/UserMention/UserMention.tsx (hunk 19 above)
Comment (Julianahlk) [reply src/components/shared/UserMention/UserMention.tsx:46]: The plan was to leave this in place for to make it easier for a separate task that would update UserMention functionality

File: src/components/shared/UserInfoPopover/UserInfo.tsx [hunk 21]
```diff
@@ -4,44 +4,26 @@ import CopyableAddress from '~shared/CopyableAddress';
...
-   * exceed the call stack (obviously) and break React
-   */
-  const { image: avatarData } =
-    useUserAvatarImageFromIPFS(avatarHash || '') || {};
   return (
     <div className={styles.container}>
       <UserAvatar
         size="s"
-        address={walletAddress}
+        address={user.walletAddress}
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/shared/UserInfoPopover/UserInfo.tsx:24]: Do you mean to remove the address prop and get the walletAddress from user inside the UserAvatar component? It seems there are still several usages of UserAvatar that haven't been ported yet, can we always assume for the cdapp that the user prop will always have the walletAddress?

File: src/components/common/ColonyHome/ColonyMembersWidget/MembersSubsection.tsx [hunk 22]
```diff
@@ -151,12 +152,13 @@ const MembersSubsection = ({
       <ul className={styles.userAvatars}>
         {(colonyWatchers as { user: User }[])
           .slice(0, avatarsDisplaySplitRules)
-          .map(({ user: { walletAddress } }) => (
-            <li className={styles.userAvatar} key={walletAddress}>
+          .map(({ user }) => (
```
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/common/ColonyHome/ColonyMembersWidget/MembersSubsection.tsx:155]: I wanted to both pass the user as a prop into UserAvatar and use the walletAddress for the key

File: src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx (hunk 15 above)
Comment (Julianahlk) [reply src/components/shared/TokenInfoPopover/TokenInfoPopoverWrapper.tsx]: @willm30  Seems like some of this was leftover from the `InfoPopover`, I simplified this and it should address your concerns now

File: src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx (hunk 8 above)
Comment (Julianahlk) [reply src/components/shared/TokenInfoPopover/TokenInfoPopover.tsx:20]: Thanks! This works 👍

File: src/components/shared/UserInfoPopover/UserReputation.tsx (hunk 18 above)
[the remark this answers is not in the document]
Comment (Julianahlk) [reply src/components/shared/UserInfoPopover/UserReputation.tsx:61]: Yes, I agree. I am happy to create any tasks for the refactoring of any areas/components you want to see improved. Just let me now what you have in mind and I will do it 😉.

**Review by rdig — APPROVED** [review]
Had a look at your latest changes. All good from my end.

You did a very good job with this PR overall. :100:
